### PR TITLE
Implement EventScrubber

### LINKED ISSUE #1897: Implement `config.event_scrubber`
### Problem Statement

Add an event scrubber as per https://github.com/getsentry/rfcs/pull/62

### Solution Brainstorm

^

### PR DESCRIPTION
**Pull request by sl0thentr0py**
Add a new `EventScrubber` class that scrubs certain potentially sensitive interfaces with a `DEFAULT_DENYLIST`. The default scrubber is automatically run if `send_default_pii = False`.
 
### Usage
```python
from sentry_sdk.scrubber import EventScrubber, DEFAULT_DENYLIST

# custom denylist
denylist = DEFAULT_DENYLIST + ["my_sensitive_var"]

sentry_sdk.init(
    # ...
    send_default_pii=False,
    event_scrubber=EventScrubber(denylist=denylist),
)
```

As outlined in https://github.com/getsentry/rfcs/blob/main/text/0062-controlling-pii-and-credentials-in-sd-ks.md
closes #1897 
docs: https://github.com/getsentry/sentry-docs/pull/6489

### CODE REVIEW

File: sentry_sdk/serializer.py [hunk 1]
```diff
@@ -254,6 +254,8 @@ def _serialize_node_impl(
         obj, is_databag, should_repr_strings, remaining_depth, remaining_breadth
     ):
         # type: (Any, Optional[bool], Optional[bool], Optional[int], Optional[int]) -> Any
+        if isinstance(obj, AnnotatedValue):
```
Comment (antonpirker) [inline sentry_sdk/serializer.py:257]: Could this check be in `_should_repr_strings()`?

**Review by antonpirker — COMMENTED** [review]
Why not just walk the whole event structure recursively and apply `scrub_dict()`?

As the implementation is now, when we (or somenone else) adds a top level key "super_secret_stuff" it will not be touched.

**Comment by antonpirker** [comment]
I get that having those seperate scrubbing functions makes it convenient for a user to derive a custom scrubber that works like the original one, but does not scrub the breadcrumbs.

**Comment by antonpirker** [comment]
But maybe add a `scrub_others` or `scrub_rest`

That does something like:
```
for key in event.keys():
    if key not in ["request", "extra", "user", ...]:
        self.scrub_dict(event[key])
```

**Comment by sl0thentr0py** [comment]
> Why not just walk the whole event structure recursively and apply scrub_dict()?

Only python has the `Event` as an opaque dict/object without interfaces, no other SDK does this, so I just went with a structure that would work for all SDKs. The recursive walking can only work in python sadly.

I think this is fine for a first version and we can see how and if users use this at all and wait for feedback.
I just wanted to have something like this as a first-class abstraction and this is a good first step.

**Comment by antonpirker** [comment]
Ok. 
How can we ensure, that if we add some top level key to the event in the future, that we do not forget to update the default EventScrubber? Because we (or whoever is than working on this) will forget about this.

File: sentry_sdk/client.py [hunk 2]
```diff
@@ -249,6 +253,11 @@ def _prepare_event(
             self.options["project_root"],
         )
 
+        if event is not None:
```
Comment (cleptric) [inline sentry_sdk/client.py:256]: Shall we run this after the `before_send` callbacks?

Comment (sl0thentr0py) [reply sentry_sdk/client.py:256]: we always said `before_send` sees what actually gets sent so let's not change that semantic

**Review by antonpirker — APPROVED** [review]

**Pull request merged**